### PR TITLE
Ticket 2: MSW learning and implementing

### PR DESCRIPTION
**Pull request by eyal0226**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,27 +1,16 @@
...
 
 function App() {
-  interface MSWRes {
-    name: string;
-    brand: string;
-    image?: string;
-    price: number;
-    id: string;
-  }
+  const [MockRes, setData] = useState<Product[]>([]);
```
Comment (bar-cohen-1) [inline src/App.tsx:7]: rename the state
the state and setter method should have the same name
find more meaningful name than "data"

File: src/App.tsx [hunk 2]
```diff
@@ -1,27 +1,16 @@
...
+  setData([]);
```
Comment (bar-cohen-1) [inline src/App.tsx]: Remove this line

File: src/mocks/data/products.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+export const Products = [
```
Comment (bar-cohen-1) [inline src/mocks/data/products.ts]: Add explicit typescript declaration here

```ts
export const Products: Product[] = [ /...
```

**Comment by bar-cohen-1** [comment]
Add index export file to `src/mocks/routes`

File: src/App.tsx [hunk 4]
```diff
@@ -1,27 +1,24 @@
...
-          <ProductCard
-            key={item.id}
-            name={item.name}
-            brand={item.brand}
-            price={item.price}
-            image={item.image}
-          />
-        ))}
+      {MockRes.map((item) => (
+        <ProductCard key={item.id} product={item} />
```
Comment (amitmarko) [inline src/App.tsx:20]:
```suggestion
      {MockRes.map((product) => (
        <ProductCard key={item.id} product={product} />
```

File: src/mocks/routes/handlers.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { rest } from "msw";
+import { Products } from "../data";
+
+export const handlers = [
+  rest.get("/products/all", (req, res, ctx) => {
+    return res(ctx.status(200), ctx.json(Products));
```
Comment (amitmarko) [inline src/mocks/routes/handlers.js]: Move to controller as we talked about.

**Review by bar-cohen-1 — APPROVED** [review]

**Review by bar-cohen-1 — APPROVED** [review]

File: src/shared/models/product.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export interface Product {
+  id: string;
+  name: string;
+  brand: string;
+  image?: string;
```
Comment (amitmarko) [inline src/shared/models/product.ts]:
```suggestion
  image: string;
```

File: src/mocks/data/products.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { Product } from "shared/models";
+
+export const Products: Product[] = [
+  {
+    name: "Air Jordan 1 Low",
+    brand: "Nike",
```
Comment (amitmarko) [inline src/mocks/data/products.ts:6]: Please create brand type Model 
`type Brand = 'Nike' | 'Adidas' | 'PUMA'`

File: src/shared/models/product.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export interface Product {
+  id: string;
+  name: string;
+  brand: string;
```
Comment (amitmarko) [inline src/shared/models/product.ts]:
```suggestion
  brand: Brand;
```

File: src/shared/models/product.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export type Brand = "Nike" | "Adidas";
```
Comment (amitmarko) [inline src/shared/models/product.ts]: `Brand` should be on brand file..

Comment (eyal0226) [reply src/shared/models/product.ts]: fixed.

File: src/App.tsx (hunk 1 above)
Comment (eyal0226) [reply src/App.tsx:7]: fixed

File: src/mocks/data/products.ts (hunk 7 above)
Comment (eyal0226) [reply src/mocks/data/products.ts:6]: fixed

File: src/shared/models/product.ts (hunk 8 above)
Comment (eyal0226) [reply src/shared/models/product.ts]: fixed

File: src/mocks/routes/handlers.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { rest } from "msw";
+import { productController } from "mocks/controllers";
+
+export const handlers = [
+  rest.get("/products/all", productController.getAll),
```
Comment (amitmarko) [inline src/mocks/routes/handlers.js]: export const apiUrl = ({ id, sourceId }: Params = {}) => ({
  prodcuts: {
    getAll: '/api/products',
    edit: `/api/secret-stores/${id}`,
   ...
  },
})


`Products: 

1. List of Products /products (GET)
2. Create Product /products (POST)
3. Delete Product /product/:id (DELETE)
4. Get Product /product/:id (GET)
5. Edit Product /product/:id (PUT/PATCH)
`

File: src/mocks/services/product.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
+import { Products } from "../data";
+import { CreateProduct, Product } from "shared/models";
+
+class ProductService {
+  getAllProducts() {
+    return Products;
+  }
+  editProductsList(editDetails: Product) {
+    let choosenProduct: Product | undefined = Products.find(
```
Comment (amitmarko) [inline src/mocks/services/product.ts]:
```suggestion
    const selectedProduct = Products.find(
```

File: src/mocks/services/product.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      (product) => editDetails.id === product.id
+    );
+    if (!choosenProduct) {
+      throw new Error("No product found");
+    }
+    Object.assign(choosenProduct, editDetails);
+    return choosenProduct;
+  }
+  deleteProduct(id: string) {
+    const newProducts = Products.filter((product) => product.id !== id);
```
Comment (amitmarko) [inline src/mocks/services/product.ts]: Add slice

File: src/mocks/services/product.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    if (newProducts.length < Products.length) {
+      return true;
+    } else {
+      throw new Error("No product found");
+    }
+  }
+  addProduct(product: CreateProduct) {
+    let new_product = {
```
Comment (amitmarko) [inline src/mocks/services/product.ts]:
```suggestion
    const newProduct = {
```

File: src/mocks/routes/handlers.js (hunk 10 above)
Comment (eyal0226) [reply src/mocks/routes/handlers.js]: fixed

File: src/mocks/services/product.ts (hunk 11 above)
Comment (eyal0226) [reply src/mocks/services/product.ts]: fixed

File: src/mocks/services/product.ts (hunk 12 above)
Comment (eyal0226) [reply src/mocks/services/product.ts]: fixed

File: src/mocks/services/product.ts (hunk 13 above)
Comment (eyal0226) [reply src/mocks/services/product.ts]: fixed

**Review by amitmarko — APPROVED** [review]
LGTM

**Pull request merged**